src/exodus/sigmawallet.cpp

```diff
@@ -91,195 +72,193 @@ uint32_t SigmaWallet::CreateNextSeed(CKeyID &seedId, uint512& seed)
 
 uint32_t SigmaWallet::GenerateSeed(CKeyID const &seedId, uint512& seed)
 {
+    LOCK(pwalletMain->cs_wallet);
     CKey key;
-    if (!pwalletMain->CCryptoKeyStore::GetKey(seedId, key)) {
+    if (!pwalletMain->GetKey(seedId, key)) {
         throw std::runtime_error(
             "Unable to retrieve generated key for mint seed. Is the wallet locked?");
     }
 
-    // HMAC-SHA512(SHA256(count), key)
-    std::array<unsigned char, CSHA256::OUTPUT_SIZE> countHash;
-
-    std::vector<unsigned char> result;
-    result.resize(CSHA512::OUTPUT_SIZE);
-
-    auto count = std::to_string(GetSeedIndex(seedId));
-    CSHA256().
-        Write(
-            reinterpret_cast<const unsigned char*>(count.data()),
-            count.size()
-        ).
-        Finalize(countHash.data());
+    // HMAC-SHA512(key, count)
+    // `count` is LE unsigned 32 bits integer
+    std::array<unsigned char, CSHA512::OUTPUT_SIZE> result;
+    auto seedIndex = GetSeedIndex(seedId);
 
-    CHMAC_SHA512(countHash.data(), sizeof(countHash)).
-        Write(key.begin(), key.size()).
+    CHMAC_SHA512(key.begin(), key.size()).
+        Write(reinterpret_cast<const unsigned char*>(&seedIndex), sizeof(seedIndex)).
         Finalize(result.data());
 
     seed = uint512(result);
 
-    return GetSeedIndex(seedId);
+    return seedIndex;
 }
 
-bool GetChildFromPath(string const &path, std::pair<uint32_t, bool> &child)
+namespace {
+
+std::uint32_t GetBIP44AddressIndex(std::string const &path)
 {
-    auto startPos = path.find_last_of('/') + 1;
-    if (startPos >= path.size()) {
-        return false;
+    auto lastSlash = path.find_last_of('/');
+    if (lastSlash == std::string::npos) {
+        throw std::runtime_error("Fail to match BIP44 path");
     }
 
-    auto childData = path.substr(startPos);
-
-    auto isHardened = childData.back() == '\'';
-    if (isHardened) {
-        childData.resize(childData.size() - 1);
+    auto child = std::stol(path.substr(lastSlash + 1));
+    if (child > std::numeric_limits<uint32_t>::max()) {
+        throw std::runtime_error("Address index is exceed limit");
     }
 
-    if (childData.empty()) {
-        return false;
-    }
+    return child;
+}
 
-    auto childPos = std::stol(childData);
+secp_primitives::Scalar GetSerialFromPublicKey(
```

Comment:
Our Exodus don't need ECDSA key to sign anything so we don't need to derive serial from ECDSA key.